### DISCUSSION TITLE
FilePortal cannot select directories or files at the same time

### DISCUSSION THREAD

**Discussion by chergert** [flatpak/xdg-desktop-portal#1419] [Ideas]
### Operating System

Fedora 40

### XDG Desktop Portal version

1.18

### Desktop Environment

GNOME

### Expected Behavior

Currently we do not have API in the portal that allows for selecting either a file or a directory. In fact, this has been a problem since the beginning when we moved to `GtkFileChooserNative` in GTK.

Apparently we did not have a bug open for this because the discussion on it predates the issue tracker for, at the time, xdg-app.

GNOME Builder allows you to select either a project directory or a file to open using the file-chooser. If the file-filter contains `inode/directory`, it dynamically switches to `SELECT_FOLDER` instead of `OPEN`. This way you can just open a project directory or, alternatively, select a project file to open (CMakeLists.txt, configure.ac, meson.build, etc etc).

We've never had a way to do this via the FilePortal and now that GTK is moving towards `GtkFileDialog` which encodes this functionality as async functions, that becomes even harder.

One thought is that we could encoded this behavior in the actual implementations, and simply change their mode based on the file-filter being `inode/directory`.

### Current Behavior

As above

### Steps to Reproduce

Try to use `org.freedesktop.portal.FileChooser.*` API to select a folder or open a file.

**Comment by chergert** [comment]
This should not be marked as a bug, but I have no idea how to change that. More of an incomplete-design we never finished before the ship sailed a decade ago.

**Comment by Mikenux** [comment]
@chergert: For anything that is not a bug, you can go to https://github.com/flatpak/xdg-desktop-portal/discussions/new/choose. The most relevant (obvious) choice for a feature request is "Idea". Maintainers can convert a bug into a "discussion".

**Comment by GeorgesStavracas** [comment]
Took the liberty to convert to a discussion

**Comment by ptomato** [comment]
I have a use case for this. In the [Inform](https://github.com/flathub/com.inform7.IDE) programming environment, users can give their Inform programs extra capabilities by installing extensions into their project. Previously, these extensions came in the form of text files, named `*.i7x`. In the upcoming version, Inform is gaining support for [extensions in the form of folders](https://github.com/ganelson/inform-evolution/blob/main/proposals/0001-extensions-with-resources.md). These will come both in the form of folders named `*.i7xd`, as well as zip archives of folders, named `*.i7xd.zip`.

I would like to be able to create a file dialog that will allow the user to pick an extension to install into their project, no matter whether it is a folder, a text file, or a zip file.

**Comment by cole-miller** [comment]
At [Zed](https://github.com/zed-industries/zed) we'd also benefit from this. Is it realistic to change the spec to support this? If so, what can I do to get the ball rolling?

**Comment by mikix** [comment]
I'll add another use case. Déjà Dup Backups would use this - when users are picking what to include or exclude from a backup, folders tend to be the priority, but I've had lots of requests to allow choosing individual files.

**Comment by kovaxis** [comment]
I also came across this, and it does look like an artificial limitation. My use case is selecting files/folders for uploading. In this scenario it doesn't really matter whether the user selects a file or a folder, both work the same.

I also noticed that rfd (a Rust library that abstracts over multiplatform file dialogues) has a feature request for file/dir dialogues, and the Linux XDG Portal backend is blocked by this. (https://github.com/PolyMeilex/rfd/issues/131)
